In [1]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand

# Random Walk

This exercise is based on Module 9.5 in *Introduction to Computational Science* by Shiflet and Shiflet.

Many physical processes are a result of an object traveling in a medium through random collisions with molecules or atoms in the medium. Examples include a pollen particle in a fluid, calcium traveling through a bone, or a photon generated in the interior of the sun traveling to the surface. One way to simulate this process is a *random walk*.

In this notebook, you will:

1. randomly select steps for the object.
1. visualize the path of the object.
1. calculate the distance traveled from its starting location.

## Background

Sir Ronald Ross (1857-1932) was a British medical doctor who received the 1902 Nobel Prize in Medicine for his work on malaria. He was not a mathematician by trade, but he was interested in calculating the distance of a mosquito-carried disease from a breeding pool. In September 1904, at the International Congress of Arts and Science in St. Louis, Ross gave a presentation on how to mathematically model the spread of mosquitoes from such a breeding pool.

Introducing the topic, he says:

"Suppose that a mosquito is born at a given point, and that during its life wanders about, to and from, to left or to right, where it wills, in search of food or mating, over a country which is uniformly attractive to it. After a time, it will die. What are the probabilities that its dead body will be found at a given distance from its birthplace? That is really the problem which governs the whole of this great subject of the prophylaxis of malaria."

He also says:

"The answer depends upon the distance which a mosquito can traverse, not during a single flight, but during its whole life; and so upon certain laws of probability, which must govern its wanderings to and fro upon the face of the earth"

Ross considered the mosquito's life to be divided into $n$ stages. During each stage, it travels a total distance $L$ in the air. He considered mosquitos traveling in all directions, but for mathematical simplicity he considered a single mosquito to travel along one axis only (i.e. one dimension). During each stage, the mosquito randomly decides *twice* to fly forward or backward a distance $L/2$, with equal probability 1/2 of flying forward or backward.

Ross did not give this process a name; however, today mathematicians and scientists call this a *one-dimensional random walk*.

Ross calculated the probability of finding a dead mosquito at a distance $L$, $2L$, $3L$, etc. for $n$ stages. As a specific example, he considered 1024 mosquitos with $n=5$ stages. The expected number of mosquitos at each possible distance from the breeding pool is shown below.

Distance from the breeding pool | Number of dead mosquitos
:---: | :---:
0 | 252
L | 420
2L | 240
3L | 90
4L | 20
5L | 2



## Exercise 1

1. Use Python to calculate the probability of finding a dead mosquito at each of the distances in the table above.
(Hint: What is the number of "successes" at each distance step? What is the denominator or total number of mosquitos?)
2. In a new text cell below, copy and paste the markdown for the above table, and add a third column that shows the probability you calculated for each distance.

In [9]:
N = 1024
N0 = 252
N1 = 420
N2 = 240
N3 = 90
N4 = 20
N5 = 2

print(f"0: {N0/N:.3f}")
print(f"1: {N1/N:.3f}")
print(f"2: {N2/N:.3f}")
print(f"3: {N3/N:.3f}")
print(f"4: {N4/N:.3f}")
print(f"5: {N5/N:.3f}")

0: 0.246
1: 0.410
2: 0.234
3: 0.088
4: 0.020
5: 0.002


Distance from the breeding pool | Number of dead mosquitos | Probability of finding mosquito
:---: | :---: | :---:
0 | 252 | 0.246
L | 420 |  0.410
2L | 240 | 0.234
3L | 90 | 0.088
4L | 20 | 0.020
5L | 2 | 0.002


## Mosquito Model

Let's define a Cartesian coordinate system with $+x$ to the right and $+y$. An object can only be at a location on the grid given by integers $(x,y)$. In other words, the location of the object given by the coordinate pair could be something like $(0,1)$ or $(-10,15)$ but not $(3.5,1)$.  A movement from one location to another is called a *displacement*.

- The mosquito will fly along the x-axis at $y=0$.
- In each stage, a mosquito will fly a total distance of $L$ by taking **two** random "steps", of distance $L/2$ to the right or to the left, with equal probability.
- A step to the right is $\Delta x = +L/2$.
- A step to the left is $\Delta x = -L/2$.
- Use $L=1$ for simplicity.

In the rest of the notebook, we will create a grid and move the mosquito randomly along the x-axis. Since we are only moving along the x-axis, we say this is a one-dimensional random walk.


## One Mosquito

Below is the code to run the random walk for a single mosquito. For each of the items below identify which line of code is responsible for that part of the walk. Copy that line(s) of code and add it to the markdown in this text cell. In markdown you can format code like \`code\`, which will format like `code`. I have done this for the first item in the list.

- Set the number of stages we will walk: ` N = 5 #total number of stages`
- Set the initial position of the mosquito: `x = 0`
- Loop over the total number of stages: `for i in range(N):`
- Randomly pick 0 or 1 to determine if we step right or left:
```
r = rand.randint(0,1)
  if r == 0:
      dx1 = L/2 #step right
   else:
      dx1 = - L/2 #step left
```
- Calculate the total distance traveled in the stage:` dx = dx1 + dx2 #total displacement`
- Update the x position after each stage:`x = x + dx #update the value of x`
- Print the x position after each stage: `print(f"Stage {i+1}, x = {x}")`

In [ ]:
N = 5 #total number of stages
L = 1 #length of two steps in the same direction

#starting position
x = 0

print(f"Stage 0, x = {x}")

#loop over the number of stages
for i in range(N):

    #step right or left
    r = rand.randint(0,1)
    if r == 0:
        dx1 = L/2 #step right
    else:
        dx1 = - L/2 #step left

    #step right or left
    r = rand.randint(0,1)
    if r == 0:
        dx2 = L/2 #step right
    else:
        dx2 = - L/2 #step left

    dx = dx1 + dx2 #total displacement
    x = x + dx #update the value of x

    print(f"Stage {i+1}, x = {x}")


Stage 0, x = 0
Stage 1, x = 0.0
Stage 2, x = -1.0
Stage 3, x = -1.0
Stage 4, x = 0.0
Stage 5, x = -1.0


## Using Python Functions

Use a Python function to do a random walk and return the final x position. I took the previous program, removed the `print` statements, and put it into a function. The variable `N` is passed to the function.


In [10]:
def randomwalk1D(N): #input N, the number of stages; returns final position
    L = 1 #length of two steps in the same direction

    #starting position
    x = 0

    #loop over the number of stages
    for i in range(N):

        #step right or left
        r = rand.randint(0,1)
        if r == 0:
            dx1 = L/2 #step right
        else:
            dx1 = - L/2 #step left

        #step right or left
        r = rand.randint(0,1)
        if r == 0:
            dx2 = L/2 #step right
        else:
            dx2 = - L/2 #step left

        dx = dx1 + dx2 #total displacement
        x = x + dx #update the value of x

    return x


To get the final x position of a mosquito, call the function and pass it the number of stages as shown below.

In [12]:
Nstages = 5

x=randomwalk1D(Nstages)
print(f"final x = {x}")

final x = -1.0


Using the above code as an example, calculate and print the distance traveled by a single mosquito after undergoing a random walk for 10 stages in the code cell below.

In [13]:
Nstages = 10

x=randomwalk1D(Nstages)
print(f"final x = {x}")

final x = 5.0


## Modeling Many Mosquitos

We want to model the *random walk* of 1024 mosquitos. We need a loop to run 1024 times. **Create a loop below and call the function `randomwalk1D`.** Run the loop 1024 times. Note that when testing, run the loop fewer times (maybe 10 times) and print each final x-position.

Hints: You can use your solution to Exercise 1 of the 02-01-random notebook as an example. How will you need to change that to run the random walk rather than roll a di?

In [15]:
Nmos = 10 #number of mosquitos
Nstages = 5

#loop
for i in range(Nmos):
  x = randomwalk1D(Nstages)
  print(f"final x = {x}")


final x = -1.0
final x = -1.0
final x = 1.0
final x = 0.0
final x = 3.0
final x = 1.0
final x = 2.0
final x = 2.0
final x = 1.0
final x = 2.0


1. After you know your program is working, copy and paste it into the cell below (don't forget to adjust the number of mosquitos up to the full 1024 and remove the print statements!).
1. Add variables to track the final position of the mosquito (i.e. `N0`, `N1`, `N2`, etc.)
1. Add code *inside the loop* to:

 - calculate the distance of the final x position of the mosquito (hint: $|x|$ can be expressed in Python code as `np.abs(x)`)
 - count the number of times each distance $L$, $2L$, $3L$... occurs by updating the variables we created in step 2. (hint: this is similar to how we tracked which circle the dart landed in last class)


In [16]:
Nmos = 1024 #number of mosquitos
Nstages = 5

# variables to track the output
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

#loop
for i in range(Nmos):
  x = randomwalk1D(Nstages)
  dist = np.abs(x)
  if dist == 0:
    N0 += 1
  if dist == 1:
    N1 += 1
  if dist == 2:
    N2 += 1
  if dist == 3:
    N3 += 1
  if dist == 4:
    N4 += 1
  if dist == 5:
    N5 += 1

print(f"N0 = {N0}")
print(f"N1 = {N1}")
print(f"N2 = {N2}")
print(f"N3 = {N3}")
print(f"N4 = {N4}")
print(f"N5 = {N5}")

N0 = 275
N1 = 401
N2 = 238
N3 = 91
N4 = 16
N5 = 3


Copy and paste your program into the cell below. Then, after the loop, calculate the percentage of mosquitos that die at each distance--this is the probability--and compare your results to the table given by Ross. (In Excerise 1 you already calcualted the probability, how can you adjust that code?)

In [29]:
Nmos = 1024 #number of mosquitos
Nstages = 5

# variables to track the output
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

#loop
for i in range(Nmos):
  x = randomwalk1D(Nstages)
  dist = np.abs(x)
  if dist == 0:
    N0 += 1
  if dist == 1:
    N1 += 1
  if dist == 2:
    N2 += 1
  if dist == 3:
    N3 += 1
  if dist == 4:
    N4 += 1
  if dist == 5:
    N5 += 1

print(f"0 | {N0} | {N0/Nmos:.3f}")
print(f"L | {N1}| {N1/Nmos:.3f}")
print(f"2L | {N2}| {N2/Nmos:.3f}")
print(f"3L | {N3}| {N3/Nmos:.3f}")
print(f"4L | {N4}| {N4/Nmos:.3f}")
print(f"5L | {N5}| {N5/Nmos:.3f}")

0 | 283 | 0.276
L | 387| 0.378
2L | 240| 0.234
3L | 87| 0.085
4L | 25| 0.024
5L | 2| 0.002


Simulated results of distance from breeding pool.

Distance from the breeding pool | Number of dead mosquitos | Simulated probability of finding mosquito
:---: | :---: | :---:
0 | 251 | 0.245
L | 421| 0.411
2L | 237| 0.231
3L | 85| 0.083
4L | 26| 0.025
5L | 4| 0.004

There is quite good agreement between our results for the distance from the breeding pool, and the results by Ross. They generally agree to within two significant figures

## Exercise 2

Run your program modeling 1024 mosquitos multiple times and observe the results each time. Remember to record your answers in narritive format (i.e. full sentances)

1. Do the percentages (i.e. probabilities) for each distance change?

2. How can you change your program in order to better estimate the expected percentage of mosquitos at each distance? Make your suggested change.

Compare your results to the table by Ross.

Our probabilities change each times, the most signifigant change is seen in the distance $L$. The better estimate the percentage we should run the model with a larger number of mosquitos.

In [30]:
Nmos = 100000 #number of mosquitos
Nstages = 5

# variables to track the output
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

#loop
for i in range(Nmos):
  x = randomwalk1D(Nstages)
  dist = np.abs(x)
  if dist == 0:
    N0 += 1
  if dist == 1:
    N1 += 1
  if dist == 2:
    N2 += 1
  if dist == 3:
    N3 += 1
  if dist == 4:
    N4 += 1
  if dist == 5:
    N5 += 1

print(f"0 | {N0} | {N0/Nmos:.3f}")
print(f"L | {N1}| {N1/Nmos:.3f}")
print(f"2L | {N2}| {N2/Nmos:.3f}")
print(f"3L | {N3}| {N3/Nmos:.3f}")
print(f"4L | {N4}| {N4/Nmos:.3f}")
print(f"5L | {N5}| {N5/Nmos:.3f}")

0 | 24788 | 0.248
L | 40645| 0.406
2L | 23646| 0.236
3L | 8791| 0.088
4L | 1944| 0.019
5L | 186| 0.002


These probabilities agree with Ross to nearly three signifigant figures.